### PR TITLE
Framebuffer formats

### PR DESCRIPTION
**Pull request by linkmauve**
Implement the remaining framebuffer formats in the OpenGL renderer, and rename the default format from RGB8 to BGR8, since that’s what it is.

### CODE REVIEW

**Comment by Alegend45** [comment]
LGTM

**Comment by Subv** [comment]
LGTM

**Comment by Normmatt** [comment]
LGTM

File: src/video_core/renderer_opengl/renderer_opengl.cpp [hunk 1]
```diff
@@ -171,6 +185,52 @@ void RendererOpenGL::InitOpenGLObjects() {
     glBindTexture(GL_TEXTURE_2D, 0);
 }
 
+void RendererOpenGL::GenerateFramebufferTexture(TextureInfo& texture,
```
Comment (yuriks) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: This function can return a `TextureInfo` by value instead.

Comment (yuriks) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: Can we rename this to something more along the lines of `ConfigureFramebufferTexture`?

File: src/video_core/renderer_opengl/renderer_opengl.cpp [hunk 2]
```diff
@@ -83,6 +81,24 @@ void RendererOpenGL::SwapBuffers() {
...
+    case GPU::Regs::PixelFormat::RGBA8:
+        return 4;
+    case GPU::Regs::PixelFormat::BGR8:
+        return 3;
+    case GPU::Regs::PixelFormat::RGB565:
+    case GPU::Regs::PixelFormat::RGB5A1:
+    case GPU::Regs::PixelFormat::RGBA4:
+        return 2;
+    default:
+        return -1;
```
Comment (yuriks) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: Make this assert false (or UNIMPLEMENTED if #500 gets merged first) instead

**Comment by yuriks** [comment]
Mostly good

File: src/citra_qt/debugger/graphics_framebuffer.h [hunk 3]
```diff
@@ -50,7 +50,7 @@ class GraphicsFramebufferWidget : public BreakPointObserverDock {
 
     enum class Format {
         RGBA8    = 0,
-        RGB8     = 1,
+        BGR8     = 1,
```
Comment (neobrain) [inline src/citra_qt/debugger/graphics_framebuffer.h]: Why?

Comment (yuriks) [inline src/citra_qt/debugger/graphics_framebuffer.h]: Because someone doesn't read PR descriptions, lol.

File: src/video_core/renderer_opengl/renderer_opengl.cpp [hunk 4]
```diff
@@ -83,6 +81,24 @@ void RendererOpenGL::SwapBuffers() {
 }
 
 /**
+ * Returns the number of bytes per pixel, or -1 if the format is unknown.
+ */
+static int BytesPerPixel(GPU::Regs::PixelFormat format) {
```
Comment (neobrain) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: This belongs to gpu.h.

File: src/video_core/renderer_opengl/renderer_opengl.cpp (hunk 1 above)
Comment (neobrain) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: Why can't it return TextureInfo by value?

Comment (yuriks) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: It takes the pre-created handle to the OpenGL texture object.

Comment (neobrain) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: Why isn't that passed as a parameter instead?

File: src/citra_qt/debugger/graphics_framebuffer.h (hunk 3 above)
Comment (neobrain) [inline src/citra_qt/debugger/graphics_framebuffer.h]: :|

@linkmauve Can you
a) explain in the commit message as well why this change was made?
b) document in the code which system the component order follows?

Comment (yuriks) [inline src/citra_qt/debugger/graphics_framebuffer.h]: The system apparently takes BGR, which is why the name's being changed. I don't see why this merits special documentation. A mention in the commit message sounds reasonable.

Comment (neobrain) [inline src/citra_qt/debugger/graphics_framebuffer.h]: Let's take RGBA8 as an example for a minute. I would say there are two options which could be the case:
* the first byte in a data stream is a red component, while the second one is blue, etc or
* the first (little-endian) word in a data stream constitutes a pixel, within which the most-significant byte constitutes the red component while the least-significant byte is the alpha component. For little-endian, this means the red component is actually the fourth byte in a data stream.

So, which one do we intend to follow? I guess the first one is intended (and is the only one which really makes sense for RGB8). Regardless, the previous code had all formats named uniformly, so there was less reason to explain this. But with this new change, my question is immediate IMO and its answer hence should be documented in the source.

File: src/video_core/renderer_opengl/renderer_opengl.cpp [hunk 5]
```diff
@@ -171,6 +185,52 @@ void RendererOpenGL::InitOpenGLObjects() {
...
+
+    case GPU::Regs::PixelFormat::RGB5A1:
+        internal_format = GL_RGBA;
+        texture.gl_format = GL_RGBA;
+        texture.gl_type = GL_UNSIGNED_SHORT_5_5_5_1;
+        break;
+
+    case GPU::Regs::PixelFormat::RGBA4:
+        internal_format = GL_RGBA;
+        texture.gl_format = GL_RGB;
```
Comment (yuriks) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: Shouldn't this be `GL_RGBA`?

File: src/citra_qt/debugger/graphics_framebuffer.h (hunk 3 above)
Comment (yuriks) [inline src/citra_qt/debugger/graphics_framebuffer.h]: So what you're saying is that, RGB8, being the only non-word or halfword type in the list, is inconsistent since it's acessed byte by byte, instead of per-word? I see your point then. There's no good way of making this un-ambiguous, since byte access only makes sense for `RGBA8` and `RGB8`, and word access only for non-`RGB8`. There's no notation that makes sense for both.

Comment (neobrain) [inline src/citra_qt/debugger/graphics_framebuffer.h]: My point is not necessarily inconsistency with regards to how things are accessed, but rather that the code doesn't explain how it comes up with the component order in the naming :)

For what it's worth, you could still say something like "read N bits from the raw byte stream" (so for RGB565, it would read 5 bits first, then 6, then 5)* - that would work for all formats. I actually think the OpenGL specification deals with the naming issue in this way. I'll look it up and check how they phrase it exactly.

* In a way this just says that the pixel is laid out in big-endian.

**Comment by yuriks** [comment]
Since #500 was merged, many games now assert on boot since they use RGBA8 by default for a frame or two. I think we should expedite fixing and merging this. @linkmauve do you have time to do this?

**Comment by linkmauve** [comment]
I have my last exam tomorrow morning, so I will be available again starting from then!

File: src/video_core/renderer_opengl/renderer_opengl.cpp (hunk 1 above)
Comment (linkmauve) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: TextureInfo is a struct containing an immutable handle to the actual OpenGL texture in the driver, as well as the current state of this texture.  It is already passed as a parameter, and I see no point in returning it while we can just modify it in-place.

File: src/citra_qt/debugger/graphics_framebuffer.h (hunk 3 above)
Comment (linkmauve) [inline src/citra_qt/debugger/graphics_framebuffer.h]: OpenGL uses byte order, so effectively big-endian, I didn’t see a reason not to change that.  Also, that convention is shared with others 3DS tools like ctrulib, with every other format in the RGB[A] order.

I will add a mention of that in the commit.

Comment (linkmauve) [inline src/citra_qt/debugger/graphics_framebuffer.h]: Actually I think I am wrong, after studying https://github.com/mtheall/screenshot_png it seems the original order was correct, as little-endian. I’m going to remove that commit.

**Comment by linkmauve** [comment]
Addressed all comments, and rebased against master. :)

**Comment by yuriks** [comment]
Looks yummy to me.

**Comment by purpasmart96** [comment]
LGTM

File: src/video_core/renderer_opengl/renderer_opengl.cpp [hunk 6]
```diff
@@ -171,6 +168,53 @@ void RendererOpenGL::InitOpenGLObjects() {
...
+
+    switch (format) {
+    case GPU::Regs::PixelFormat::RGBA8:
+        internal_format = GL_RGBA;
+        texture.gl_format = GL_RGBA;
+        texture.gl_type = GL_UNSIGNED_INT_8_8_8_8;
+        break;
+
+    case GPU::Regs::PixelFormat::RGB8:
+        // This format uses BGR since GL_UNSIGNED_BYTE specifies byte-order, instead of native-endian.
```
Comment (neobrain) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: Thanks for clearing this up, but... how can you speak about endianness in a 24-bit format? If OpenGL defines pixel formats via the byte order, wouldn't it make more sense for us to follow that convention as well and rename **all** pixel format enum elements accordingly?

File: src/video_core/renderer_opengl/renderer_opengl.cpp [hunk 7]
```diff
@@ -171,6 +168,53 @@ void RendererOpenGL::InitOpenGLObjects() {
...
+        texture.gl_format = GL_RGBA;
+        texture.gl_type = GL_UNSIGNED_SHORT_5_5_5_1;
+        break;
+
+    case GPU::Regs::PixelFormat::RGBA4:
+        internal_format = GL_RGBA;
+        texture.gl_format = GL_RGBA;
+        texture.gl_type = GL_UNSIGNED_SHORT_4_4_4_4;
+        break;
+    }
```
Comment (neobrain) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: No default case here. Is that intentional and/or do we check for unknown formats elsewhere already?

Comment (yuriks) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: Good point. Whatever we do it should be consistent. (See https://github.com/citra-emu/citra/pull/539/files#diff-7f006a0ada2bd1e8eedc003ec49e0e7cR78 )

File: src/video_core/renderer_opengl/renderer_opengl.h [hunk 8]
```diff
@@ -43,9 +43,14 @@ class RendererOpenGL : public RendererBase {
         GLuint handle;
         GLsizei width;
         GLsizei height;
+        GPU::Regs::PixelFormat format;
+        GLenum gl_format;
+        GLenum gl_type;
     };
 
     void InitOpenGLObjects();
+    void ConfigureFramebufferTexture(TextureInfo& texture,
```
Comment (neobrain) [inline src/video_core/renderer_opengl/renderer_opengl.h]: This can be made static.

File: src/video_core/renderer_opengl/renderer_opengl.cpp (hunk 7 above)
Comment (linkmauve) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: I think we’ll need some hwtests in order to know how to behave when 5 < format < 8, but for the time being I added UNIMPLEMENTED() there.

File: src/video_core/renderer_opengl/renderer_opengl.cpp (hunk 6 above)
Comment (linkmauve) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: So ABGR8 and BARG4 and such? :/
I think we could just say that everything is in reverse-order (compared to byte-order).

Comment (neobrain) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: Yeah, reverse byte-order sounds fine.

Comment (linkmauve) [inline src/video_core/renderer_opengl/renderer_opengl.cpp]: Fixed the note to reflect that.

**Comment by neobrain** [comment]
Could you set in stone how we decide to call the PixelFormat enum elements (reverse component-order) at the place its defined at (gpu.h)? These things are an easy source of confusion and hence should be mentioned explicitly instead of assumed to be obvious.

**Comment by linkmauve** [comment]
Here, added that comment (and expanded the other one a bit).

**Comment by bunnei** [comment]
LGTM
